config/config_generator/main.go

```diff
@@ -0,0 +1,42 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See License.txt for license information.
+
+package main
+
+import (
+	"encoding/json"
+	"flag"
+	"fmt"
+	"github.com/mattermost/mattermost-server/config"
+	"io/ioutil"
+	"os"
+)
+
+func usage() {
+	fmt.Fprintf(os.Stderr, "usage: config_generator [output_file]\n")
+	flag.PrintDefaults()
+	os.Exit(2)
+}
+
+func main() {
+	flag.Usage = usage
+	flag.Parse()
+
+	args := flag.Args()
+	if len(args) < 1 {
+		fmt.Println("Input file is missing.")
+		os.Exit(1)
+	}
+
+	configStore, err := config.NewFileStore("config.json", true)
```

Comment:
Unfortunately, this won't work for two reasons:
* If someone uses this with an existing `config.json`, they'll generate a `default.json` with perhaps non-default values. This won't be a problem in CI, but would be confusing otherwise.
* In addition to `SetDefaults`, the file store also generates some /other/ initial values that we /don't/ want in `default.json` (various keys are randomly generated).

Instead, I think we just need to create the file direclty from the defaults, e.g.:
```go
cfg := &model.Config{}
cfg.SetDefaults()

b, err := marshalConfig(cfg)
if err != nil {
    return errors.Wrap(err, "failed to serialize")
}

err = ioutil.WriteFile(args[0], b, 0644)
if err != nil {
    return errors.Wrap(err, "failed to write file")
}
```

Note that `marshalConfig` is not exported from `config`, so perhaps this generate code can just live as `default_generate.go` directly in the `config` package?

Reply:
I am confused again :) If we are just creating a new `default.json` without taking `config.json` into account - how will we recognize if it's a new/updated config?
We can avoid using file store, but I don't see how we can avoid try to load and upgrade existing `config.json`.

Also, about your note: we cannot move the tool to default_generate.go , because it'll create a GO file with package name that differs from 'main' and go generate will not be able to execute it, so I'll leave it where it is, but replace marshalConfig() with a regular call to json.Marshal (because that's all that function does)

Reply:
Recognizing if it's a new/updated config happens purely inside the context of `SetDefaults`. The server does this automatically whenever it reads the config (whether from a file or from the database).

*This* code is purely about generating the default `config.json`, so it doesn't need to start from an existing file at all. In fact, we must ensure it does not -- otherwise it won't be truly "default". Only the CI server will be using your config generator, and it never needs to "upgrade" an existing file: it's generating our *.tar.gz files that contain the pristine default config file.

Re: `main`, you're correct again: we won't be able to mix package names as I proposed. I suppose another option would be to use the memory store -- it always starts from the defaults. Then maybe we don't need to export anything new?

Reply:
Actually, I guess we'd still need a `MarshalConfig` or to just write `json.MarshalIndent` directly. Let's keep your export so that we avoid duplicate code (and can ensure these implementations stay in-sync).